plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostStream.hs

```diff
@@ -0,0 +1,128 @@
+    deriving stock (Show)
+
+-- TODO: (# CostingInteger, (# (# #) | CostStream #) #)?
+-- | Uncons an element from a 'CostStream' and return the rest of the stream, if not empty.
+unconsCost :: CostStream -> (CostingInteger, Maybe CostStream)
+unconsCost (CostLast cost)       = (cost, Nothing)
+unconsCost (CostCons cost costs) = (cost, Just costs)
+{-# INLINE unconsCost #-}
+
+-- | Cons an element to a 'CostStream', if given any. Otherwise create a new 'CostStream' using
+-- 'CostLast'.
+reconsCost :: CostingInteger -> Maybe CostStream -> CostStream
+reconsCost cost = maybe (CostLast cost) (CostCons cost)
+{-# INLINE reconsCost #-}
+
+{- Note [Global local functions]
+Normally when defining a helper function one would put it into a @where@ or a @let@ block.
+However if the enclosing function gets inlined, then the definition of the helper one gets inlined
+too, which when happens in multiple places can create serious GHC Core bloat, making it really hard
+to analyze the generated code. Hence in some cases we optimize for lower amounts of produced GHC
+Core by turning some helper functions into global ones.
+
+This doesn't work as well when the helper function captures a variables bound by the enclosing one,
+so we leave such helper functions local. We could probably create a global helper and a local
+function within it instead, but so far it doesn't appear as those capturing helpers actually get
+duplicated in the generated Core.
+-}
+
+-- See Note [Global local functions].
+sumCostStreamGo :: CostingInteger -> CostStream -> CostingInteger
+sumCostStreamGo !acc (CostLast cost)       = acc + cost
+sumCostStreamGo !acc (CostCons cost costs) = sumCostStreamGo (acc + cost) costs
+
+-- | Add up all the costs in a 'CostStream'.
+sumCostStream :: CostStream -> CostingInteger
+sumCostStream (CostLast cost0)        = cost0
+sumCostStream (CostCons cost0 costs0) = sumCostStreamGo cost0 costs0
+{-# INLINE sumCostStream #-}
+
+-- See Note [Global local functions].
+-- | Map a function over a 'CostStream'.
+mapCostStream :: (CostingInteger -> CostingInteger) -> CostStream -> CostStream
+-- See Note [Single-element streams]
+mapCostStream f (CostLast cost0)        = CostLast (f cost0)
+mapCostStream f (CostCons cost0 costs0) = CostCons (f cost0) $ go costs0 where
+    go :: CostStream -> CostStream
+    go (CostLast cost)       = CostLast (f cost)
+    go (CostCons cost costs) = CostCons (f cost) $ go costs
+{-# INLINE mapCostStream #-}
+
+-- See Note [Global local functions].
+addCostStreamGo :: CostStream -> CostStream -> CostStream
+addCostStreamGo (CostLast costL)        costsR = CostCons costL costsR
+addCostStreamGo (CostCons costL costsL) costsR = CostCons costL $ addCostStreamGo costsR costsL
+
+-- | Add two streams by interleaving their elements (as opposed to draining out one of the streams
+-- before starting to take elements from the other one). No particular reason to prefer
+-- interleaving over draining out one of the streams first.
```

Comment:
You could see it as some kind of "fairness"

Reply:
I don't think I can claim fairness, since we use this function to interleave three streams too and that results in twice more elements drawn from the first stream compared to the other two. Plus I just don't feel like we care about fairness there: if favoring one stream over another can result in an issue, then one could just use the same argument twice and make fairness irrelevant.